source/common/http/conn_manager_impl.cc

```diff
@@ -1475,13 +1475,18 @@ ConnectionManagerImpl::FilterManager::commonDecodePrefix(
   return std::next(filter->entry());
 }
 
-void ConnectionManagerImpl::startDrainSequence() {
+void ConnectionManagerImpl::startDrainSequence(bool wait_for_timeout) {
   ASSERT(drain_state_ == DrainState::NotDraining);
   drain_state_ = DrainState::Draining;
   codec_->shutdownNotice();
-  drain_timer_ = read_callbacks_->connection().dispatcher().createTimer(
-      [this]() -> void { onDrainTimeout(); });
-  drain_timer_->enableTimer(config_.drainTimeout());
+  if (wait_for_timeout) {
```

Comment:
Why would we not do the timer in all cases? This timer is designed to avoid races in which the client races with the goaway and then has the streams reset. Given that we will need to wait until all existing streams drain anyway I don't see any reason to not do the normal drain sequence here?

Reply:
The goal of this PR is to make H2 clients more responsive to overload conditions. We actually *want* to stop the clients from making any more requests, even if that means racing with incoming requests. If we wait for the timer, then there is a lag between when the proxy is overloaded and when it actually takes action.

Reply:
The default drain timer is I think 5s, and in general we try to make things graceful, similar to what happens with H1. I'm very skeptical that it's worth the extra code complexity and non-graceful behavior for this special case when the default is only around 5s?

Reply:
Agreed, I've switched this to drain in the normal way, sending the shutdown notice first and then the GOAWAY after the delay.